lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
 
-    {test_module, pending, finished_tests} =
+    {test_module, test_results} =
       receive do
-        {^module_pid, :module_finished, test_module, failed_tests, finished_tests} ->
-          Process.demonitor(module_ref, [:flush])
-          {test_module, failed_tests, finished_tests}
+        {^module_pid, :module_finished, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
+          else
+            process_failure(config, test_module)
+            Process.demonitor(module_ref, [:flush])
+            {test_module, tests}
+          end
+
+        {^module_pid, :max_failures_reached, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
+          else
+            Process.demonitor(module_ref, [:flush])
+            {test_module, tests}
+          end
 
         {:DOWN, ^module_ref, :process, ^module_pid, error} ->
-          test_module = %{test_module | state: failed({:EXIT, module_pid}, error, [])}
-          {test_module, [], []}
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            {test_module, %{pending: [], finished: []}}
+          else
+            test_module = %{test_module | state: failed({:EXIT, module_pid}, error, [])}
+            process_failure(config, test_module)
+            {test_module, %{pending: [], finished: []}}
+          end
       end
 
-    {exec_on_exit(test_module, module_pid, timeout), pending, finished_tests}
+    test_module = exec_on_exit(test_module, module_pid, timeout)
+    {test_module, test_results}
+  end
+
+  defp spawn_module_monitor(config, test_module, parent_pid, tests) do
+    spawn_monitor(fn ->
+      module_pid = self()
+      register(module_pid)
+
+      case exec_module_setup(test_module) do
+        {:ok, _test_module, context} ->
+          # max_failures can be reached during the execution of test_module,
+          # so we keep track of which tests are not executed and which ones are finished.
+          tests_run = run_tests(config, tests, context)
+
+          send(
+            parent_pid,
+            {module_pid, :module_finished, test_module,
+             %{pending: [], finished: tests_run.finished}}
+          )
+
+        {:error, test_module} ->
+          invalid_tests = Enum.map(tests, &%{&1 | state: {:invalid, test_module}})
+
+          send(
+            parent_pid,
+            {module_pid, :module_finished, test_module, %{pending: invalid_tests, finished: []}}
+          )
+      end
+
+      exit(:shutdown)
+    end)
+  end
+
+  # returns a map in the shape of
+  # %{:not_executed => list(), :finished => list()}
+  defp run_tests(config, tests, context) do
+    Enum.map(tests, fn test ->
```

Comment:
Minor nitpick:

```elixir
tests
|> Enum.map(tests, &run_test(config, &1, context)) # or  fn test -> run_test(config, test, context)
|> Enum.group_by(&elem(&1, 0), &elem(&1, 1))
|> Map.put_new(:not_executed, [])
|> Map.put_new(:finished, [])
```

Reply:
We should not return a map here though. In fact, we should stop running the other tests as soon as we get the first `not_executed`.